specs/agents/tracing-spans-limit.md

```diff
@@ -0,0 +1,55 @@
+[Agent spec home](README.md) > [Handling huge traces](tracing-spans-handling-huge-traces.md) > [Hard limit on number of spans to collect](tracing-spans-limit.md)
+
+## Hard limit on number of spans to collect
+
+This is the last line of defense that comes with the highest amount of data loss.
+This strategy MUST be implemented by all agents.
+Ideally, the other mechanisms limit the amount of spans enough so that the hard limit does not kick in.
+
+### Configuration option `transaction_max_spans`
+
+Limits the amount of spans that are recorded per transaction.
+
+This is helpful in cases where a transaction creates a very high amount of spans (e.g. thousands of SQL queries).
+
+Setting an upper limit will prevent overloading the agent and the APM server with too much work for such edge cases.
+
+|                |          |
+|----------------|----------|
+| Type           | `integer`|
+| Default        | `500`    |
+| Dynamic        | `true`   |
+
+### Implementation
+
+Before creating a span,
+agents must determine whether creating that span would exceed the span limit.
+The limit is reached when the total number of created spans minus the number of dropped spans is greater or equals to the max number of spans.
+In other words, the limit is reached if this condition is true:
+
+    span_count.total - span_count.dropped >= transaction_max_spans
+
+The `span_count.total` counter is not part of the intake API,
+but it helps agents to determine whether the limit has been reached.
+It reflects the total amount of started spans within a transaction.
+
+To ensure consistent behavior within one transaction,
+the `transaction_max_spans` option should be read once on transaction start.
+Even if the option is changed via remote config during the lifetime of a transaction,
+the value that has been read at the start of the transaction should be used.
+
+Note that it's not enough to just consider this condition on span start:
+
+    span_count.sent >= transaction_max_spans
```

Comment:
The issue is not a race condition but the fact that `span_count.sent` can only be incremented on span end, after it has been determined whether or not this span should be sent or dropped.

(That reminded me that after a span has been compressed we either have to increment `span_count.dropped` or decrement `span_count.total`)

However, that has the issue of potentially dropping more spans than needed. Consider this scenario:
- transaction_max_spans=100
- Start 200 concurrent spans
- The first 100 are dropped due to `exit_span_min_duration`
- -> We don't report any spans at all as the second 100 of the spans are dropped due to `transaction_max_spans`

An alternative would be to check `span_count.sent` both at the start AND at the end of the span:
- On start: `dropped = atomic_read( span_count.sent ) >= transaction_max_spans`
- On end: `dropped = atomic_read( span_count.sent ) >= transaction_max_spans || atomic_get_and_increment( span_count.sent ) >= transaction_max_spans`

This may be the best of both worlds where agents can determine at the start of the span whether it would definitely get over the limit so that they can omit collecting expensive data for that span. But it would never under-shoot the limit.

Reply:
> The issue is not a race condition but the fact that `span_count.sent` can only be incremented on span end, after it has been determined whether or not this span should be sent or dropped.

Yes, I agree - eagerly allocating "a space in span count" on span's start has a cons of potentially of keeping space for spans that are going to be sent and thus discarding spans that could have been sent otherwise.  

> This may be the best of both worlds where agents can determine at the start of the span whether it would definitely get over the limit so that they can omit collecting expensive data for that span. But it would never under-shoot the limit.

Yes, I agree - the amount of potentially "started but then dropped" spans should be no more than max amount of spans running simultaneously.

OTOH I wonder if this optimization is even worth the effort considering that we need to do some work even for dropped spans (for example including them in metrics) so it seems that the savings will be considerable only for spans that need to do work for span's instance if it's to be sent out but not for spans metrics - such as collecting HTTP request/response headers/body. Is it correct?

Reply:
That is correct. The spans have to be collected anyway but things like the SQL summarization or the capturing of request bodies for Elasticsearch spans can be skipped if it's known that the current spans won't be reported at all. However, the timing info and other metadata, such as the destination still have to be collected, even for spans that exceed the `transaction_max_spans` limit.

Reply:
I have revised this section. Could you have another look?